### PR TITLE
fixed: Chunk size logic for file cache (hopefully fixes #16975)

### LINKED ISSUE #16975: Video stops playing from a samba share after a couple of seconds
### Describe the bug
Here is a clear and concise description of what the problem is:

- I do a fresh installation of kodi 18.5 on a rasperry pi 4 with libreelec
- I have a samba share to a drive connected on my router (192.168.1.1) defined in kodi
- Playing video's from that share works fine
- Then I create an advancedsettings.xml file with only this:

```
<advancedsettings>
  <cache>
    <buffermode>1</buffermode>
  </cache>
</advancedsettings>
```

I restart kodi and play the video again. After a couple of seconds the video just stops playing.

When I set buffermode to 0 then the videos play fine again.

When I set buffermode again to 1 and go to kodi Services, SMB client, Maximum protocol version and set this to SMBv1 then the video plays fine again.

So there seems to be a problem with buffering local file system on and samba share using with version 2 or greater. Since samba 1 should no longer be used this becomes a problem.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

The video should play without a problem with buffering local filesystem on and samba 2 or greater

## Actual Behavior

The video stops after a couple of seconds

### To Reproduce

I already provided the steps above

### Debuglog

The debuglog can be found here:

I provide 3 debuglogs. 

I do 3 times the same thing. Paying video smb://192.168.1.1/TomAndJerry/Tom en Jerry Show/Zogenaamd getrouwd.avi

- One with buffermode = 0 (playing ok):  https://paste.kodi.tv/bodecumuwo.kodi
- One with buffermode = 1 (stopping after couple secs): https://paste.kodi.tv/wopideteju.kodi
- One with buffermode = 1, max smbv1, playing ok: https://paste.kodi.tv/vecuhucigo.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] Linux
 - [ ] OSX
 - [*] Raspberry-Pi 4
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Libreelec
 - Kodi version: 18.5

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by arnova**
Also adds the option to make the default(!) chunksize configurable for filecache. Additionally it contains a few other (minor) tweaks.

### CODE REVIEW

**Comment by fritsch** [comment]
Thanks much. Looks good from first glance @MilhouseVH could you create a build for the user in the linked bugreport?

**Comment by arnova** [comment]
@fritsch : Please note that I chose to limit the as.xml configurable chunk size between 256 bytes and 1 MB. Do those sound sane to you?

**Comment by fritsch** [comment]
@arnova If people want to shoot themselves in the shoes :-) we should allow them. It's okay :-)

**Comment by MilhouseVH** [comment]
RPi2 LibreELEC/Kodi 19 build here:

http://milhouse.libreelec.tv/builds/testing/RPi2/LibreELEC-RPi2.arm-9.80-Milhouse-20191216141812-%231215b-gb28a93e.tar

Using the following cache settings in advancedsettings.xml:
```
<cache>
  <buffermode>1</buffermode>
</cache>
```

I see...

`smb://`:
```
2019-12-16 14:26:56.496 T:1097   DEBUG: CFileCache::Open - Source chunk size is 65536, setting cache chunk size to 65536
```
`nfs://`:
```
2019-12-16 14:29:16.056 T:1121   DEBUG: CFileCache::Open - Source chunk size is 32768, setting cache chunk size to 32768
```
YouTube:
```
2019-12-16 14:34:06.507 T:1448   DEBUG: CFileCache::Open - Source chunk size is 0, setting cache chunk size to 131072
```

**Comment by fritsch** [comment]
@MilhouseVH Thanks. Anything you realized with "problematic" files?

**Comment by MilhouseVH** [comment]
I'm using wired ethernet and the `smb://` and `nfs://` videos played fine, as did the YouTube video until Kodi crashed towards the end of the video - not sure if that's related to this change or just a general issue/problem. Will see if I can reproduce (unfortunately I didn't manage to capture a full crash log, but it was due to a seg fault):
```
Using host libthread_db library "/usr/lib/libthread_db.so.1".
Core was generated by `/usr/lib/kodi/kodi.bin --standalone -fs'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x730d24a8 in ?? () from /usr/lib/libmmal_vc_client.so
[Current thread is 1 (Thread 0x660ff330 (LWP 1437))]
```

**Comment by MilhouseVH** [comment]
Seems to have been a one-off as I played the same video through to completion without crashing.

File: xbmc/filesystem/CacheStrategy.cpp [hunk 1]
```diff
@@ -119,7 +119,7 @@ int CSimpleFileCache::WriteToCache(const char *pBuffer, size_t iSize)
   size_t written = 0;
   while (iSize > 0)
   {
-    const ssize_t lastWritten = m_cacheFileWrite->Write(pBuffer, (iSize > SSIZE_MAX) ? SSIZE_MAX : iSize);
+    const ssize_t lastWritten = m_cacheFileWrite->Write(pBuffer, std::min(iSize, (size_t)SSIZE_MAX));
```
Comment (Rechi) [inline xbmc/filesystem/CacheStrategy.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting


> New code has to use C++ style casts and not older C style casts. [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#122-casts

File: xbmc/filesystem/CacheStrategy.cpp [hunk 2]
```diff
@@ -145,14 +145,14 @@ int CSimpleFileCache::ReadFromCache(char *pBuffer, size_t iMaxSize)
 {
   int64_t iAvailable = GetAvailableRead();
   if ( iAvailable <= 0 )
-    return m_bEndOfInput? 0 : CACHE_RC_WOULD_BLOCK;
+    return m_bEndOfInput ? 0 : CACHE_RC_WOULD_BLOCK;
 
-  size_t toRead = ((int64_t)iMaxSize > iAvailable) ? (size_t)iAvailable : iMaxSize;
+  size_t toRead = std::min(iMaxSize, (size_t)iAvailable);
```
Comment (Rechi) [inline xbmc/filesystem/CacheStrategy.cpp]: > New code has to use C++ style casts and not older C style casts. [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#122-casts

File: xbmc/filesystem/CacheStrategy.cpp [hunk 3]
```diff
@@ -145,14 +145,14 @@ int CSimpleFileCache::ReadFromCache(char *pBuffer, size_t iMaxSize)
...
 
   size_t readBytes = 0;
   while (toRead > 0)
   {
-    const ssize_t lastRead = m_cacheFileRead->Read(pBuffer, (toRead > SSIZE_MAX) ? SSIZE_MAX : toRead);
+    const ssize_t lastRead = m_cacheFileRead->Read(pBuffer, std::min(toRead, (size_t)SSIZE_MAX));
```
Comment (Rechi) [inline xbmc/filesystem/CacheStrategy.cpp]: > New code has to use C++ style casts and not older C style casts. [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#122-casts

File: xbmc/filesystem/File.cpp [hunk 4]
```diff
@@ -1037,7 +1043,7 @@ ssize_t CFile::LoadFile(const CURL& file, auto_buffer& outputBuffer)
   if (filesize > (int64_t)max_file_size)
     return 0; /* file is too large for this function */
 
-  size_t chunksize = (filesize > 0) ? (size_t)(filesize + 1) : (size_t) GetChunkSize(GetChunkSize(), min_chunk_size);
+  size_t chunksize = (filesize > 0) ? (size_t)(filesize + 1) : (size_t) DetermineChunkSize(GetChunkSize(), min_chunk_size);
```
Comment (Rechi) [inline xbmc/filesystem/File.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting


> New code has to use C++ style casts and not older C style casts. [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#122-casts

File: xbmc/filesystem/File.cpp [hunk 5]
```diff
@@ -1096,7 +1102,7 @@ void CFileStreamBuffer::Attach(IFile *file)
 {
   m_file = file;
 
-  m_frontsize = CFile::GetChunkSize(m_file->GetChunkSize(), 64*1024);
+  m_frontsize = CFile::DetermineChunkSize(m_file->GetChunkSize(), 64*1024);
```
Comment (Rechi) [inline xbmc/filesystem/File.cpp]: > Conventional operators have to be surrounded by one space on each side.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#34-whitespace

File: xbmc/filesystem/FileCache.cpp [hunk 6]
```diff
@@ -135,7 +133,11 @@ bool CFileCache::Open(const CURL& url)
 
   // check if source can seek
   m_seekPossible = m_source.IoControl(IOCTRL_SEEK_POSSIBLE, NULL);
-  m_chunkSize = CFile::GetChunkSize(m_source.GetChunkSize(), READ_CACHE_CHUNK_SIZE);
+
+  // Determine the best chunk size we can use
+  m_chunkSize = CFile::DetermineChunkSize(m_source.GetChunkSize(), CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_cacheChunkSize);
```
Comment (Rechi) [inline xbmc/filesystem/FileCache.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/filesystem/FileCache.cpp [hunk 7]
```diff
@@ -135,7 +133,11 @@ bool CFileCache::Open(const CURL& url)
...
+  CLog::Log(LOGDEBUG, "CFileCache::Open - Source chunk size is %i, setting cache chunk size to %i", m_source.GetChunkSize(), m_chunkSize);
```
Comment (Rechi) [inline xbmc/filesystem/FileCache.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/filesystem/FileCache.cpp [hunk 8]
```diff
@@ -239,7 +241,7 @@ void CFileCache::Process()
         m_nSeekResult = m_source.Seek(cacheMaxPos, SEEK_SET);
         if (m_nSeekResult != cacheMaxPos)
         {
-          CLog::Log(LOGERROR,"CFileCache::Process - Error %d seeking. Seek returned %" PRId64, (int)GetLastError(), m_nSeekResult);
+          CLog::Log(LOGERROR, "CFileCache::Process - Error %d seeking. Seek returned %" PRId64, (int)GetLastError(), m_nSeekResult);
```
Comment (Rechi) [inline xbmc/filesystem/FileCache.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting


> New code has to use C++ style casts and not older C style casts. [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#122-casts

File: xbmc/filesystem/FileCache.cpp [hunk 9]
```diff
@@ -255,6 +257,7 @@ void CFileCache::Process()
         m_nSeekResult = m_seekPos;
         if (bCompleteReset)
         {
+          CLog::Log(LOGDEBUG, "CFileCache::Process - Cache completely reset for seek to position %" PRId64, m_seekPos);
```
Comment (Rechi) [inline xbmc/filesystem/FileCache.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

**Comment by fritsch** [comment]
@arnova I can send you a branch for the formatting changes if you want.

**Comment by fritsch** [comment]
Here: https://github.com/fritsch/xbmc/tree/cache_chunk_fix (if you fetch it reset hard it) -> Changes:https://paste.ubuntu.com/p/nMFKXBbjmc/

**Comment by arnova** [comment]
@fritsch: Thanks

**Comment by arnova** [comment]
@rechi / @fritsch : Should be fully updated now.

**Comment by peno64** [comment]
Does this mean that this change will be in the nightly build and thus tomorrow available to test?

**Comment by fritsch** [comment]
When I press the button, yes.

**Comment by fritsch** [comment]
jenkins build this please

**Comment by fritsch** [comment]
UWP build failure unrelated. @arnova if you are happy with the state, please push :-)

File: xbmc/filesystem/File.cpp [hunk 10]
```diff
@@ -1037,7 +1043,8 @@ ssize_t CFile::LoadFile(const CURL& file, auto_buffer& outputBuffer)
   if (filesize > (int64_t)max_file_size)
     return 0; /* file is too large for this function */
 
-  size_t chunksize = (filesize > 0) ? (size_t)(filesize + 1) : (size_t) GetChunkSize(GetChunkSize(), min_chunk_size);
+  size_t chunksize = (filesize > 0) ? static_cast<size_t>(filesize + 1)
+                                    : static_cast<size_t>(DetermineChunkSize(GetChunkSize(), min_chunk_size));
```
Comment (Rechi) [inline xbmc/filesystem/File.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/filesystem/CacheStrategy.cpp [hunk 11]
```diff
@@ -145,14 +146,14 @@ int CSimpleFileCache::ReadFromCache(char *pBuffer, size_t iMaxSize)
...
+    return m_bEndOfInput ? 0 : CACHE_RC_WOULD_BLOCK;
 
-  size_t toRead = ((int64_t)iMaxSize > iAvailable) ? (size_t)iAvailable : iMaxSize;
+  size_t toRead = std::min(iMaxSize, static_cast<size_t>(iAvailable));
 
   size_t readBytes = 0;
   while (toRead > 0)
   {
-    const ssize_t lastRead = m_cacheFileRead->Read(pBuffer, (toRead > SSIZE_MAX) ? SSIZE_MAX : toRead);
+    const ssize_t lastRead = m_cacheFileRead->Read(pBuffer, std::min(toRead, static_cast<size_t>(SSIZE_MAX)));
```
Comment (Rechi) [inline xbmc/filesystem/CacheStrategy.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

File: xbmc/filesystem/File.cpp [hunk 12]
```diff
@@ -414,6 +414,12 @@ bool CFile::OpenForWrite(const CURL& file, bool bOverWrite)
   return false;
 }
 
+int CFile::DetermineChunkSize(const int srcChunkSize, const int reqChunkSize)
+{
+  // Determine cache chunk size: if source chunk size is bigger than 1 use source chunk size else use requested chunk size
```
Comment (Rechi) [inline xbmc/filesystem/File.cpp]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

**Comment by arnova** [comment]
Again I've updated it for formatting. But to be honest I'm getting a bit sick of this 100-column limit bullshit. It's not like it's improving readability in every single location. Furthermore it sometimes takes more time to adjust it for that crap than fixing the actual bug itself.

**Comment by arnova** [comment]
jenkins build this please

**Comment by fritsch** [comment]
jenkins build this please

**Comment by fritsch** [comment]
Failures all not related.

**Comment by ronie** [comment]
@arnova i'm not able to connect to our repository since the PR got merged.

```
2019-12-26 16:43:52.233 T:6988   DEBUG: CRepositoryUpdateJob[repository.xbmc.org] checking for updates.
2019-12-26 16:43:52.235 T:6988   DEBUG: CFileCache::Open - opening <addons/matrix/addons.xml.gz> using cache
2019-12-26 16:43:52.235 T:6988   DEBUG: CurlFile::Open(0x7f9e100087b0) http://mirrors.kodi.tv/addons/matrix/addons.xml.gz?sha256
2019-12-26 16:43:52.235 T:6988    INFO: easy_acquire - Created session to http://mirrors.kodi.tv
2019-12-26 16:43:52.286 T:6988   DEBUG: CFileCache::Open - Source chunk size is 0, setting cache chunk size to 131072
2019-12-26 16:43:52.286 T:7003   DEBUG: Thread FileCache start, auto delete: false
2019-12-26 16:43:52.332 T:6967   DEBUG: ------ Window Init (DialogBusy.xml) ------
2019-12-26 16:44:02.312 T:7003   DEBUG: Thread FileCache 140316145338112 terminating
2019-12-26 16:44:02.313 T:6988   DEBUG: CurlFile::Open(0x7f9e19f36c20) http://mirrors.kodi.tv/addons/matrix/addons.xml.gz
2019-12-26 16:44:02.484 T:6988   DEBUG: CCurlFile::Open - effective URL: <https://ftp2.nluug.nl/mediaplayer/xbmc/addons/matrix/addons.xml.gz>
2019-12-26 16:44:02.773 T:6988   ERROR: CRepository: http://mirrors.kodi.tv/addons/matrix/addons.xml.gz index has wrong digest 24cbf87e98de1668089f56e8910df24b3c8facdaeb8b6c748ee527d586df4d29, expected:
```

full log: https://paste.kodi.tv/bopenofeca.kodi

tested on linux / clean install
it worked fine before this PR was merged.

**Comment by fritsch** [comment]
```
2019-12-26 16:43:52.286 T:6988   DEBUG: CFileCache::Open - Source chunk size is 0, setting cache chunk size to 131072
```

The problem seems, that we now default up to 128 KB, but in our front we only allocate 64 KB. @arnova is that true for the case input has 0 as chunksize?

```
m_frontsize = CFile::DetermineChunkSize(m_file->GetChunkSize(), 64 * 1024);
```
GetChunkSize: 0
 64 * 1024 < 131072 :-(
should at least be able to store the chunk we read, right?

**Comment by arnova** [comment]
@fritsch: ah yes I think you are right but will have a closer look tomorrow. currently not at home due to 2nd christmas day dinner.

**Comment by fritsch** [comment]
@MilhouseVH According to your release site, the fix is in. As the user tested your version before, which was working fine - any idea what channged so that it now fails?

On a second note the "-" error is something else in my opinion and we should not track new bugs and fixed bugs in same issues.

**Comment by peno64** [comment]
From https://github.com/xbmc/xbmc/issues/16975
I understand that https://github.com/xbmc/xbmc/pull/17042 was merged into the master branch, but https://github.com/xbmc/xbmc/pull/17046 wasn't.
So maybe this is the reason that the issue is not fixed?

**Comment by MilhouseVH** [comment]
@fritsch this PR is in my [Generic (x86_64)](https://forum.kodi.tv/showthread.php?tid=343069) builds (as of `#1225`) but it's not in [RPi builds](https://forum.kodi.tv/showthread.php?tid=343068) `#1225` and `#1226` as there's currently a conflict with @popcornmix's newclock5 branch caused by this PR and given the time of year I'm not inclined to pester anyone to rebase their branch so that I can build  and instead I've just reverted this PR for the time being. It's mentioned in the [RPi release notes](https://forum.kodi.tv/showthread.php?tid=343068&pid=2910874#pid2910874). When it is included in the RPi builds it will be mentioned in the "build highlights".

As for what has been tested I doubt anyone tested accessing repositories with this PR applied - I'm pretty sure they only tested Samba as that's what this PR seemed to be addressing.

Is the RPi2 test build [`#1215b`](https://github.com/xbmc/xbmc/pull/17042#issuecomment-566089567) failing for @peno64 when accessing repositories? I just tested a Skylake NUC with my latest `#1226` build which includes this PR and I couldn't see any of the "digest" related failures reported by @ronie (I also tested a clean install, and again no digest-related errors: http://ix.io/25A1 only #17049). In all cases I clicked "Check for updates" from the blade menu.

**Comment by arnova** [comment]
> From #16975
> I understand that #17042 was merged into the master branch, but #17046 wasn't.
> So maybe this is the reason that the issue is not fixed?

#17046 is not relevant to your issue.

**Comment by arnova** [comment]
> @arnova i'm not able to connect to our repository since the PR got merged.
> 
> ```
> 2019-12-26 16:43:52.233 T:6988   DEBUG: CRepositoryUpdateJob[repository.xbmc.org] checking for updates.
> 2019-12-26 16:43:52.235 T:6988   DEBUG: CFileCache::Open - opening <addons/matrix/addons.xml.gz> using cache
> 2019-12-26 16:43:52.235 T:6988   DEBUG: CurlFile::Open(0x7f9e100087b0) http://mirrors.kodi.tv/addons/matrix/addons.xml.gz?sha256
> 2019-12-26 16:43:52.235 T:6988    INFO: easy_acquire - Created session to http://mirrors.kodi.tv
> 2019-12-26 16:43:52.286 T:6988   DEBUG: CFileCache::Open - Source chunk size is 0, setting cache chunk size to 131072
> 2019-12-26 16:43:52.286 T:7003   DEBUG: Thread FileCache start, auto delete: false
> 2019-12-26 16:43:52.332 T:6967   DEBUG: ------ Window Init (DialogBusy.xml) ------
> 2019-12-26 16:44:02.312 T:7003   DEBUG: Thread FileCache 140316145338112 terminating
> 2019-12-26 16:44:02.313 T:6988   DEBUG: CurlFile::Open(0x7f9e19f36c20) http://mirrors.kodi.tv/addons/matrix/addons.xml.gz
> 2019-12-26 16:44:02.484 T:6988   DEBUG: CCurlFile::Open - effective URL: <https://ftp2.nluug.nl/mediaplayer/xbmc/addons/matrix/addons.xml.gz>
> 2019-12-26 16:44:02.773 T:6988   ERROR: CRepository: http://mirrors.kodi.tv/addons/matrix/addons.xml.gz index has wrong digest 24cbf87e98de1668089f56e8910df24b3c8facdaeb8b6c748ee527d586df4d29, expected:
> ```
> 
> full log: https://paste.kodi.tv/bopenofeca.kodi
> 
> tested on linux / clean install
> it worked fine before this PR was merged.

@ronie : Are you 100% it's this PR causing it? I can't wrap my head around as to why this is happening. Do you happen to know what the offending digest is? Is it the one we obtain from http or the one we calculate from addons.xml.gz ?

**Comment by arnova** [comment]
It's something in CFileCache that broke. I think it's a bug that has always been there but has been masked until the recent changes. Will look into it further.

**Comment by arnova** [comment]
See #17081

**Comment by peno64** [comment]
@MilhouseVH :
The RPi2 test build #1215b did solve the problem. However there is another problem in that build that generates a crash when playing a video. See https://github.com/xbmc/xbmc/issues/16975#issuecomment-566699979 and the comment after it.

**Comment by fritsch** [comment]
Don't tunnel bugs in bugs. The maintainer already told you: "If so, that is a bug in current nightlies and not related to the samba/file cache issue."

**Comment by peno64** [comment]
@fritsch Sure I know. That other bug is why I wanted to install the latest nightly build hoping that that crash bug was also solved.

**Comment by MilhouseVH** [comment]
newclock5 has been rebased so this PR will be included in tonight's `#1227` RPi builds.

**Comment by peno64** [comment]
#16975 (Video stops playing from a samba share after a couple of seconds) is indeed fixed with the latest @MilhouseVH build #1227

Thank you all very much for helping in this.

Since this is/was really a bug so will this be downgraded also to Leia? I guess it will still take quite some time before Matrix will be released...

**Comment by DaveTBlake** [comment]
@arnova are you looking to backport this for v18.6 release?

**Comment by arnova** [comment]
> @arnova are you looking to backport this for v18.6 release?

Well, I'm not sure. The code is sensitive and it already caused regressions before.

**Pull request merged**